File: benchmarks/tsb/bench_memory_usage.ts
/**
 * Benchmark: seriesMemoryUsage / dataFrameMemoryUsage — memory estimation.
 * Outputs JSON: {"function": "memory_usage", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, DataFrame, seriesMemoryUsage, dataFrameMemoryUsage } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 50;

const numSeries = new Series(Array.from({ length: SIZE }, (_, i) => i * 1.0));
const strSeries = new Series(Array.from({ length: SIZE }, (_, i) => `label_${i % 100}`));

const df = new DataFrame({
  a: Array.from({ length: SIZE }, (_, i) => i * 1.0),
  b: Array.from({ length: SIZE }, (_, i) => i * 2.0),
  c: Array.from({ length: SIZE }, (_, i) => `cat_${i % 50}`),
  d: Array.from({ length: SIZE }, (_, i) => i % 2 === 0),
});

for (let i = 0; i < WARMUP; i++) {
  seriesMemoryUsage(numSeries);
  seriesMemoryUsage(strSeries, { deep: true });
  dataFrameMemoryUsage(df);
  dataFrameMemoryUsage(df, { deep: true });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  seriesMemoryUsage(numSeries);
  seriesMemoryUsage(strSeries, { deep: true });
  dataFrameMemoryUsage(df);
  dataFrameMemoryUsage(df, { deep: true });
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "memory_usage",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_memory_usage.py
"""
Benchmark: Series.memory_usage / DataFrame.memory_usage — memory estimation.
Outputs JSON: {"function": "memory_usage", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 50

num_series = pd.Series([i * 1.0 for i in range(SIZE)])
str_series = pd.Series([f"label_{i % 100}" for i in range(SIZE)])
df = pd.DataFrame({
    "a": [i * 1.0 for i in range(SIZE)],
    "b": [i * 2.0 for i in range(SIZE)],
    "c": [f"cat_{i % 50}" for i in range(SIZE)],
    "d": [i % 2 == 0 for i in range(SIZE)],
})

for _ in range(WARMUP):
    num_series.memory_usage()
    str_series.memory_usage(deep=True)
    df.memory_usage()
    df.memory_usage(deep=True)

start = time.perf_counter()
for _ in range(ITERATIONS):
    num_series.memory_usage()
    str_series.memory_usage(deep=True)
    df.memory_usage()
    df.memory_usage(deep=True)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "memory_usage",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
